Fixes 151: introspect repo from kafka message

avisiedo opened the pull request:
- Add launch.json to debug easier from vscode.
- Add kafka-introspect command to consume kafka message to introspect repositories.
- Add initial kafka support.
- Generate code for the kafka message schemas: `make build` regenerate `.json` and `.gen.go` files when the schema change.
- Update dependencies.
- Update kafka make rules to use `kafkacat` to produce a test message.
- 

TODO:
* [x] Fix test failures.
* [x] Update documentation.
* [x] Clean-up code.
* [x] Try ephemeral default pool pool
* [x] Try ephemeral managed-kafka pool
* [x] Try ephemeral real-managed-kafka pool
* [x] Log information about the received message.
* [x] Fix multi arch base image for kafka broker and zookeeper.
* [x] Add unit tests.
* [x] Add comments to public function and methods
* [x] Move kafkaconfigmap configuration in an external file (not sure the reason I wrote this item; I don't see right now the need). The reason was to have a better control on the client library parameters; the intention is not provide all the parameters, but externalize them allow to quickly update config, restart infra, and try, if some client parameter is necessary to be customized. Added card for this particular item to address in the future: https://issues.redhat.com/browse/HMSCONTENT-236

jlsherrill commented:
https://issues.redhat.com/browse/HMSCONTENT-151

avisiedo commented:
First success run in ephemeral environment:

```raw
{"level":"debug","time":"2022-09-28T08:15:48Z","message":"OnMessage was called; Key=871a77ca-f216-4a76-af30-2d76e1da04fe"}
{"level":"debug","time":"2022-09-28T08:15:48Z","message":"Introspecting http://mirror.stream.centos.org/9-stream/AppStream/x86_64/os/"}
{"level":"debug","time":"2022-09-28T08:17:27Z","message":"IntrospectionUrl returned 13106 new packages"}
```

avisiedo commented:
Similar result in pool `managed-kafka` in ephemeral environment after fix some typo:

```raw
{"level":"debug","time":"2022-09-28T09:12:45Z","message":"OnMessage was called; Key=5b707a6c-c4fa-41f0-b8a8-859da236f6d3"}
{"level":"debug","time":"2022-09-28T09:12:45Z","message":"Introspecting http://mirror.stream.centos.org/9-stream/AppStream/x86_64/os/"}
{"level":"debug","time":"2022-09-28T09:14:37Z","message":"IntrospectionUrl returned 13106 new packages"}
```

avisiedo commented on .golangci.yaml:
```diff
@@ -5,6 +5,9 @@ linters:
     - gofmt
     - whitespace
     - govet
+    - misspell
+    - forcetypeassert
+    - gci
```
Aded this three to catch hints raised by the pipeline (I would add more as I get more linter hints)

avisiedo commented on .vscode/launch.json:
```diff
@@ -0,0 +1,80 @@
+{
```
Maybe I have to delete this; it is the debug launchers I use for Visual Studio Code. Would you like to keep it or delete it?

jlsherrill replied:
@Andrewgdewar you're the only other person that uses vscode i think, your thoughts?

avisiedo commented on pkg/handler/repositories.go:
```diff
@@ -97,10 +126,67 @@ func (rh *RepositoryHandler) createRepository(c echo.Context) error {
+	}
+
+	// Fill topic
+	topic := schema.TopicIntrospect
```
BUG! this should use the real topic name; add a mapping based into clowder configuration; by default the mapping will match the same topic so it is easier to deal into the developing environment (local workstation and ephemeral environment).

avisiedo commented on pkg/event/consumer.go:
```diff
@@ -0,0 +1,216 @@
+				break
+			}
+
+			if err = validateMessage(schemas, msg); err != nil {
```
TODO Add informative log message that display: the kafka topic (source of the message), consumer group-id, key and X-Rh-Insights-Request-Id head (the last one to track along the system); that would be useful for troubleshooting.

jlsherrill commented:
* We'll need to work with bulkCreate and update as well (if url changes)
* When testing with the local UI, i get an error when trying to send the message:  
```
4:24PM WRN error producing kafka message: expected a value for 'X-Rh-Insights-Request-Id' http header
```

I'm guessing that in this particular case that header isn't present.  We should probably not make it required

rverdile commented on pkg/event/schema/schemas.go:
```diff
@@ -0,0 +1,155 @@
+	return s.prepareParseErrorList(parseErrs)
+}
+
+func (s *Schema) Validate(data interface{}) error {
```
Is this method being called anywhere? I don't see it.

avisiedo replied:
thanks! I should call this from the producer function!

avisiedo replied:
https://issues.redhat.com/browse/HMSCONTENT-222

rverdile commented on pkg/event/utils.go:
```diff
@@ -0,0 +1,107 @@
+	return "", fmt.Errorf("Header not found: %s", key)
+}
+
+func FilterByHeaderPredicate(ctx context.Context, header string, filterVals ...string) KafkaMessagePredicate {
```
Is this still being used somewhere? I can't find it being called.

avisiedo replied:
I have deleted it; not used.

rverdile commented:
There's a lot of "TODO" comments, are you gonna address all of them on this PR?

avisiedo commented:
thanks @jlsherrill ! I have updated the code to produce the messages when `bulk_create` is called; making the changes for when the url is updated too.

avisiedo commented:
Rebased and fix introspectURL invokation; it force the introspection when the message is consumed.

rverdile commented on pkg/handler/repositories.go:
```diff
@@ -222,3 +275,91 @@ func (rh *RepositoryHandler) deleteRepository(c echo.Context) error {
+	if input == nil {
+		return nil, fmt.Errorf("input cannot be nil")
+	}
+	output := &message.IntrospectRequestMessage{
```
I know for this PR it is, but is an `IntrospectRequestMessage` the only message we would send to Kafka in the future? I'm not sure how, but I wonder if this method and the `produceMessage` method should somehow handle messages more generically.

rverdile replied:
Maybe something like, you take the `RepositoryResponse` and then pass it to a method from the `event` (or similar) package that figures out what to do with the response. Kinda like the dao layer but in this case it's kafka.

avisiedo replied:
I don't forget this; thinking about it.

avisiedo replied:
I have put the adapter code at pkg.event.adapter ; I have tried to provide some semantic through the usage of the package name; Every adapter for the event message will be into an interface related with the message (idiomatic, not a direct relationship of `has ...` or `is ....`).

rverdile commented on pkg/handler/repositories.go:
```diff
@@ -222,3 +275,91 @@ func (rh *RepositoryHandler) deleteRepository(c echo.Context) error {
+	return output, nil
+}
+
+func (rh *RepositoryHandler) bulkProduceMessages(c echo.Context, response []api.RepositoryBulkCreateResponse) error {
```
Similar to my previous comment, I wonder if the kafka related methods in here should be in their own package. Then from here we just called something like `event.ProduceIntrospectMessage(msg)`, similar to how we interact with the dao layer.

avisiedo replied:
Yep! I agree! I don't like the place where I put that code either! I have to move to `pkg/event`

avisiedo replied:
moved some logic to pkg/event/producer and pkg/event/adapter; I have to update pkg/event/README.md to provide more documentation about this; in short:
- 1 interface per specific message producer; all of them use the common code for a generic message.
- The interface is registered with the http handler that will use it.
- From the handler the specific message producer is available and it is called previous call to the adapter to translate to the message format.

rverdile commented on pkg/event/message/headers.go:
```diff
@@ -0,0 +1,46 @@
+	HdrTypeIntrospect = "Introspect"
+)
+
+func (h *Header) Set(key EventHeaderKey, value string) {
+	var m map[EventHeaderKey]string = *h
+	m[key] = value
+}
```
Are the Header methods here (Set, Get, GetAll) used anywhere? I can't references find them.

avisiedo replied:
thanks! I have to clean-up this!

avisiedo commented:
/retest

rverdile commented on kafka/Dockerfile:
```diff
@@ -1,19 +1,29 @@
 
-ENV KAFKA_HOME "/opt/kafka"
+ENV KAFKA_VERSION=3.3.0
+ENV KAFKA_SITE_URL=https://dlcdn.apache.org/kafka/${KAFKA_VERSION}/kafka_2.13-${KAFKA_VERSION}.tgz
```
I'm unable to run `make kafka-build` and it seems like it could be because this URL is invalid. If I curl this manually, I get a 404. Is it just me?

avisiedo replied:
no! it happened to me during coding this!! And I needed to figure out a way to avoid this happens, other way, this will happens everytime a new versioin is released, I completely forgot it.  :+1: :+1: 

Things that cross my mind:
- the value should passed as a build argument instead of a environment variable.
- Out of the Dockerfile, the right version still needs to be retrieved; I would like to avoid to extract the information by parsing the download page (https://kafka.apache.org/downloads), or (https://dlcdn.apache.org/kafka/).

So far, I will update the version from `3.3.0` to `3.3.1`; but I have to change this to make it dynamic.

jlsherrill replied:
ah yeah, i did run into this and just updated it locally to 3.3.1.  I should have mentioned this :)  Shame that kafka doesn't include old versions.

jlsherrill replied:
Another option instead of having every user build it themselves, is we host a built version in quay.

avisiedo replied:
update: I have added a hack by using a python script to retrieve the version, but it could fails when the html layout change; I retrieve the first h3 tag so far.

avisiedo commented on cmd/kafka-introspect/main.go:
```diff
@@ -0,0 +1,17 @@
+		panic(err)
+	}
+	handler := handler.NewIntrospectHandler(db.DB)
+	event.Start(cfg, handler)
```
Addressed in a new card to be considered.

jlsherrill commented on pkg/config/event.go:
```diff
@@ -0,0 +1,68 @@
+		}
+	} else {
+		// TODO Review, probably clean-up this else
+		// If clowder is not present, set defaults to local configuration
```
is this TODO still relevant? seems the local config is being used?

avisiedo replied:
thanks! removing the todo comment; the else is necessary.

jlsherrill commented on pkg/event/config.go:
```diff
@@ -0,0 +1,70 @@
+			Ms int
+		}
+	}
+}
```
since this is more config related, shouldn't it be in ./config?

jlsherrill replied:
i should say, i think all the config structs should probably live in ./config/

avisiedo replied:
Thanks and sorry for the delay! the reason I moved this part of the struct was a circular dependency; I don't think the solution I have provided is the right one to remove that circular dependency, but I have to learn more about how to break the cycles of the dependency graph. I am open to suggestions on this; I recognize this will need refactor.

jlsherrill commented on pkg/event/config.go:
```diff
@@ -0,0 +1,70 @@
+// 	Capath                string
+// 	MessageSendMaxRetries int
+// 	RetryBackoffMs        int
+// }
```
can this be deleted?

jlsherrill commented on pkg/event/consumer.go:
```diff
@@ -0,0 +1,238 @@
+
+// TODO Refactor to make this check dynamic so it does not need
+//      to be modified if more different messages are added
+func isValidEvent(event string) bool {
```
eh, i think we can worry about that refactor when we have a new event type we're listening for

jlsherrill commented on pkg/event/schema/introspectRequest.message.yaml:
```diff
@@ -0,0 +1,36 @@
+    #   /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(? ...
+required:
+  - uuid
+  - url
```
why is there a json and yaml version of the same thing?  (or is there some difference?)

avisiedo replied:
I try to clarify below:

- If I remove the `.json` files, the tool yaml2json will be required in container build time, and building time (which will impact into the pipeline too), requiring the dependency to be installed.
- why is this happening? `make build` has a dependency on `make gen-event-messages`; this second rule has a general rule that generate the json version, if it does not exist, or if the yaml version is updated; to do that it uses yaml2json tool, which is a python tool.
- the json version is embedded into the golang code.
- the json version is required because the schema is read in json format by the library `https://github.com/qri-io/jsonschema`.

Because of all the above I left the json version into the repository even when it can be re-generated; it is a sequence of linked consequences :face_with_spiral_eyes:

I have to review `README.md`, and if the dependency for development on `yaml2json` is not mentioned, I have to add it there; this dependency is the same to use json2yaml

avisiedo replied:
update: I have added the requirement to `README.md` file.

jlsherrill commented:
Overall i this is shaping up.  I think just some cleanup and resolving the TODOs  (either via jira cards, in which case we should remove the TODO comment, deciding they aren't worth doing, or actually doing them).

avisiedo commented on mk/go-rules.mk:
```diff
@@ -54,3 +58,7 @@ test-ci: ## Run tests for ci
 
+# Regenerate code when message schema change
+$(shell find "$(EVENT_MESSAGE_DIR)" -type f -name '*.go'): $(SCHEMA_YAML_FILES)
+	$(MAKE) gen-event-messages
```
when we do `make build`, indirectly by the previous rule which add the dependencies between the binaries and all the sources; it happens when the schema file is updated, the source code date is older and make will execute the rule. yes, it is a bit tricky.

avisiedo commented:
/retest

swadeley commented on README.md:
````diff
@@ -44,9 +45,63 @@ $ cp ./configs/config.yaml.example ./configs/config.yaml
+  $ make kafka-topic-consume # Use the first topic at KAFKA_TOPICS list
+  ```
+
+> There are other make rules that could be helpful, just
````
```suggestion
> There are other make rules that could be helpful,
```

swadeley commented on README.md:
````diff
@@ -44,9 +45,63 @@ $ cp ./configs/config.yaml.example ./configs/config.yaml
+  ```
+
+> There are other make rules that could be helpful, just
+> run `make help` to check them.
````
```suggestion
> run `make help` to list them.
```

swadeley commented on README.md:
```diff
@@ -64,10 +119,11 @@ $ make run
 
 ###
+
 Hit the api:
```
```suggestion
Hit the API:
```

swadeley commented on mk/gen-event-messages.mk:
```diff
@@ -0,0 +1,45 @@
+##
+# This file add rules to generate code from the
```
```suggestion
# This file adds rules to generate code from the
```

swadeley commented on mk/go-rules.mk:
```diff
@@ -27,7 +31,7 @@ clean: ## Clean binaries and testbin generated
 
 .PHONY: run
-run: build ## Run the service locally
+run: build ## Run the service locally (it could requires kafka-introspection)
```
```suggestion
run: build ## Run the service locally (it might require kafka-introspection)
```

swadeley commented on mk/go-rules.mk:
```diff
@@ -54,3 +58,7 @@ test-ci: ## Run tests for ci
 # so any change is detected for the build rule
 $(patsubst cmd/%,$(GO_OUTPUT)/%,$(wildcard cmd/*)): $(shell find $(PROJECT_DIR)/cmd -type f -name '*.go') $(shell find $(PROJECT_DIR)/pkg -type f -name '*.go')
 
+# Regenerate code when message schema change
```
```suggestion
# Regenerate code when message schema changes
```

swadeley commented on mk/kafka.mk:
```diff
@@ -1,23 +1,35 @@
+# KAFKA_TOPICS ?= platform.content-sources.introspect
+KAFKA_GROUP_ID ?= content-sources
 
 # The Kafka configuration directory that will be bind inside the containers
```
```suggestion
# The Kafka configuration directory that will be bound inside the containers
```

swadeley commented on mk/kafka.mk:
```diff
@@ -1,23 +1,35 @@
 
 # The Kafka configuration directory that will be bind inside the containers
 KAFKA_CONFIG_DIR ?= $(PROJECT_DIR)/kafka/config
 # The Kafka data directory that will be bind inside the containers
```
```suggestion
# The Kafka data directory that will be bound inside the containers
```

swadeley commented on mk/variables.mk:
```diff
@@ -12,9 +12,23 @@
+# The name of application
+APP_NAME=content-sources
+
+# The name for this component (for instance: backend frontend)
```
```suggestion
# The name for this component (e.g.: backend, frontend)
```

swadeley commented on mk/variables.mk:
```diff
@@ -69,8 +79,27 @@ DOCKER_IMAGE_TAG ?= $(shell git rev-parse --short HEAD)
+# a topic with different group-id will receive a copy of the
+# message. In our scenario, any replica of the consumer want
+# only one message to be processed, so we only use unique
+# group id, so far
```
```suggestion
# group ID.
```
so far ? "At the moment", "At present" , but not much point to say that if there is no info about when this will change.

avisiedo replied:
thanks mate! I have updated to `at the moment`.

swadeley commented on pkg/event/schema/introspectRequest.message.yaml:
```diff
@@ -0,0 +1,36 @@
+    type: string
+    format: uuid
+  url:
+    description: The base url for the repository to be introspected
```
```suggestion
    description: The base URL for the repository to be introspected
```

swadeley reviewed (commented):
some suggestions

avisiedo commented on pkg/event/README.md:
```diff
@@ -0,0 +1,64 @@
+
+* **adapter**: It contains all the port in and port out adapters.
+  Every adapter is directly related with an event message, but the
+  one for the kafka headers, and it own as many methods as transformations
```
I mean: each adapter own as many methods as transformations. Would be more correct the below? I think I explained really bad in the original description.

```raw
* **adapter**: It contains all the port in and port out adapters.
  Every adapter is directly related with an event message, (but the
  one for the kafka headers that is associated with the headers).
  Each adapter has as many methods as transformations
  for the port in or port out flows. The external dependencies would be
  centralized here, the rest of the package use functions, structs and
  interfaces declared at `pkg/event`.
```

avisiedo replied:
@swadeley please let me know your advice on this.

avisiedo commented:
Several updates:

* Updated the pending comments.
* Updated the pending reviews.
* Reviewed the todos; some of them updated, other translated into new cards to be considered.

Thanks for all your comments and reviews: @rverdile @swadeley @jlsherrill

rverdile commented on pkg/event/handler/introspect_test.go:
```diff
@@ -0,0 +1,124 @@
+	// When a gormDb connector is passed
+	// gormDb, mock = getDatabase()
+	gormDb, _ = getDatabase()
+	// db, mock, err = sqlmock.New()
+	// require.NotNil(t, db)
+	// defer db.Close()
+	// require.NoError(t, err)
+	// require.NotNil(t, mock)
+
+	// dialector := postgres.New(postgres.Config{
+	// 	DSN:                  "sqlmock_db_0",
+	// 	DriverName:           "postgres",
+	// 	Conn:                 db,
+	// 	PreferSimpleProtocol: true,
+	// })
+	// gormDb, err = gorm.Open(dialector, &gorm.Config{})
+	// require.NoError(t, err)
+	// require.NotNil(t, gormDb)
```
was this meant to be deleted?

avisiedo replied:
thanks mate! removed!

rverdile commented on pkg/event/handler/introspect_test.go:
```diff
@@ -0,0 +1,124 @@
+			},
+			Expected: fmt.Errorf("Key: '' Error:Field validation for '' failed on the 'required' tag"),
+		},
+		// {
+		// 	Name: "Error introspecting url",
+		// 	Given: &kafka.Message{
+		// 		Value: []byte(`{"url":"https://example.test","uuid":"6742a4c0-0fe5-4abc-9037-bfbe57d3bcb5"}`),
+		// 	},
+		// 	Expected: fmt.Errorf("Error validating url:"),
+		// },
+		// {
+		// 	Name: "Success introspection",
+		// 	Given: &kafka.Message{
+		// 		Value: []byte(`{"url":"https://example.test","uuid":"6742a4c0-0fe5-4abc-9037-bfbe57d3bcb5"}`),
+		// 	},
+		// 	Expected: fmt.Errorf("Error validating url:"),
+		// },
+	}
```
was this meant to be deleted?

avisiedo replied:
thanks mate! removed!

swadeley commented on README.md:
```diff
@@ -7,7 +7,9 @@ Content Sources is an application for storing information about external content
 
 ### Requirements:
+
 1. podman installed or docker installed (and running)
```
```suggestion
1. podman or docker installed and running.
```

swadeley commented on pkg/event/README.md:
````diff
@@ -0,0 +1,92 @@
+└── message
+```
+
+* **adapter**: It contains all the port-in and port-out adapters.
````
```suggestion
* **adapter**: It contains all the input and output adapters.
```

swadeley commented on pkg/event/README.md:
```diff
@@ -0,0 +1,92 @@
+* **adapter**: It contains all the port-in and port-out adapters.
+  Every adapter is directly related with an event message, (but the
+  one for the kafka headers), and it has as many methods as transformations
+  for the port in or port out adapters. The external dependencies are
```
```suggestion
  for the input and output adapters. The external dependencies are
```

swadeley commented on pkg/event/README.md:
```diff
@@ -0,0 +1,92 @@
+  The `schema.go` file define structs and validation methods which
+  will enable validating a message before producing it, and before
+  it is consumed by the handler.
+* **message**: Auto-generated code which represent the golang
```
```suggestion
* **message**: Auto-generated code which represents the golang
```

swadeley commented on pkg/event/README.md:
```diff
@@ -0,0 +1,92 @@
+## Adding a new event message
+
+* Define the message schema at `pkg/event/schema` (yaml file).
+* Add the new schema content as a local
```
```suggestion
* Add the new schema content as a local.
```
Try to be consistent in your use of punctuation.

swadeley commented on pkg/event/README.md:
```diff
@@ -0,0 +1,92 @@
+
+## Debugging producer
+
+Currently producer is launched at the end of some http api handler.
```
```suggestion
Currently producer is launched at the end of some HTTP API handler.
```

swadeley commented on pkg/event/adapter/introspect.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/content-services/content-sources-backend/pkg/event/message"
+)
+
+// IntroespectRequestPortIn scope the IntrospectRequestPortIn adapter
```
```suggestion
// IntrospectRequestPortIn scope the IntrospectRequestPortIn adapter
```

swadeley commented on pkg/event/adapter/introspect.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+// IntroespectRequestPortIn scope the IntrospectRequestPortIn adapter
+//   so it will include any conversion to/from message.IntrospectRequestMessage
```
```suggestion
//   so it will include any conversion to or from message.IntrospectRequestMessage
```

swadeley commented on pkg/event/adapter/introspect.go:
```diff
@@ -0,0 +1,72 @@
+	return output, nil
+}
+
+// FromRepositoryRequest convert an api.RepositoryRequest plus an uuid into a message.IntrospectRequestMessage
```
```suggestion
// FromRepositoryRequest convert an api.RepositoryRequest plus a UUID into a message.IntrospectRequestMessage
```

swadeley commented on pkg/event/adapter/introspect.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+// FromRepositoryRequest convert an api.RepositoryRequest plus an uuid into a message.IntrospectRequestMessage
+// Return a message.IntrospectRequestMessage and nil error when everythin goes well, else
```
Was this fixed: when everythin <-- missing g

swadeley commented on pkg/event/config.go:
```diff
@@ -0,0 +1,54 @@
+package event
+
+// The kafka configuration that is used for produce and
```
```suggestion
// The kafka configuration that is used to produce and
```

avisiedo commented:
update: rebased with the last changes from the `main` branch

avisiedo commented:
Last todo in this PR is addressed on this card: https://issues.redhat.com/browse/HMSCONTENT-236 

Checking the todo item.

jlsherrill commented:
[test]

jlsherrill commented on deployments/deployment.yaml:
```diff
@@ -13,7 +14,56 @@ objects:
+          topicName: platform.content-sources.introspect
       deployments:
+        - name: kafka-consumer
+          minReplicas: 1
```
should this be 3?

avisiedo replied:
:+1: yes, it should be 3 thanks

jlsherrill commented:
Question there about the replicas, and then i think we can go ahead and merge this.
